exporter/opentelemetry-exporter-prometheus/src/opentelemetry/exporter/prometheus/_mapping.py

```diff
@@ -0,0 +1,135 @@
+    "1": "",
+    "%": "percent",
+}
+# Similar to _UNIT_MAPPINGS, but for "per" unit denominator.
+# Example: s => per second (singular)
+# Copied from https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/80317ce83ed87a2dff0c316bb939afbfaa823d5e/pkg/translator/prometheus/normalize_name.go#L58
+_PER_UNIT_MAPPINGS = {
+    "s": "second",
+    "m": "minute",
+    "h": "hour",
+    "d": "day",
+    "w": "week",
+    "mo": "month",
+    "y": "year",
+}
+
+
+def sanitize_full_name(name: str) -> str:
+    """sanitize the given metric name according to Prometheus rule, including sanitizing
+    leading digits
+
+    https://github.com/open-telemetry/opentelemetry-specification/blob/v1.33.0/specification/compatibility/prometheus_and_openmetrics.md#metric-metadata-1
+    """
+    # Leading number special case
+    if name and name[0].isdigit():
+        name = "_" + name[1:]
+    return _sanitize_name(name)
+
+
+def _sanitize_name(name: str) -> str:
+    """sanitize the given metric name according to Prometheus rule, but does not handle
+    sanitizing a leading digit."""
+    return _SANITIZE_NAME_RE.sub("_", name)
+
+
+def sanitize_attribute(key: str) -> str:
+    """sanitize the given metric attribute key according to Prometheus rule.
+
+    https://github.com/open-telemetry/opentelemetry-specification/blob/v1.33.0/specification/compatibility/prometheus_and_openmetrics.md#metric-attributes
+    """
+    # Leading number special case
+    if key and key[0].isdigit():
+        key = "_" + key[1:]
+    return _SANITIZE_ATTRIBUTE_KEY_RE.sub("_", key)
+
+
+def map_unit(unit: str) -> str:
+    """Maps unit to common prometheus metric names if available and sanitizes any invalid
+    characters
+
+    See:
+    - https://github.com/open-telemetry/opentelemetry-specification/blob/v1.33.0/specification/compatibility/prometheus_and_openmetrics.md#metric-metadata-1
+    - https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.101.0/pkg/translator/prometheus/normalize_name.go#L108
+    """
+    # remove curly brace UCUM annotations
+    unit = _UCUM_ANNOTATION_CURLY.sub("", unit)
+
+    if unit in _UNIT_MAPPINGS:
+        return _UNIT_MAPPINGS[unit]
+
+    # replace "/" with "per" units like m/s -> meters_per_second
+    ratio_unit_subparts = unit.split("/", maxsplit=1)
+    if len(ratio_unit_subparts) == 2:
+        bottom = _sanitize_name(ratio_unit_subparts[1])
+        if bottom:
+            top = _sanitize_name(ratio_unit_subparts[0])
+            top = _UNIT_MAPPINGS.get(top, top)
+            bottom = _PER_UNIT_MAPPINGS.get(bottom, bottom)
+            return f"{top}_per_{bottom}" if top else f"per_{bottom}"
+
+    return (
+        # since units end up as a metric name suffix, they must be sanitized
+        _sanitize_name(unit)
+        # strip surrounding "_" chars since it will lead to consecutive underscores in the
+        # metric name
+        .strip("_")
```

Comment:
> The resulting unit SHOULD be added to the metric as [UNIT metadata](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#metricfamily) and as a suffix to the metric name unless the metric name already contains the unit, or the unit MUST be omitted.

Just to clarify, we will be checking the existence of a SANITIZED and valid unit name in a sanitized and valid metric name to determine whether or not to omit the appending of the suffix correct?

Reply:
Good question, I haven't implemented that extra check yet. Since we are using the prometheus client to put together the pieces and not building the whole name ourselves, it's a little tricky.

That said I don't think "unless the metric name already contains the unit" intended for a substring match check, but rather looking for an existing `_` separated subpart containing the unit. So stripping is probably OK. I'll give it a shot

Reply:
> Since we are using the prometheus client to put together the pieces and not building the whole name ourselves, it's a little tricky.

Actually I just checked and the prometheus client actually already handles this for us. Added some test cases in 39a21a31f69a61448e4a307fcfa9e15daf3f25d3. Hope that addresses your concern